librad/src/net/protocol.rs

```diff
@@ -116,6 +121,53 @@ pub struct Protocol<S, A> {
 
     connections: Arc<Mutex<HashMap<PeerId, quic::Connection>>>,
     subscribers: Fanout<ProtocolEvent<A>>,
+
+    ref_count: Arc<AtomicUsize>,
+}
+
+impl<S, A> Clone for Protocol<S, A>
+where
+    S: Clone,
+    A: Clone,
+{
+    fn clone(&self) -> Self {
+        const MAX_REFCOUNT: usize = (isize::MAX) as usize;
+
+        let old_count = self.ref_count.fetch_add(1, atomic::Ordering::Relaxed);
+        // See source docs of `std::sync::Arc`
+        if old_count > MAX_REFCOUNT {
+            eprintln!("Fatal: max refcount on gossip::Protocol exceeded");
+            core::intrinsics::abort()
+        }
+
+        Self {
+            gossip: self.gossip.clone(),
+            git: self.git.clone(),
+            endpoint: self.endpoint.clone(),
+            connections: self.connections.clone(),
+            subscribers: self.subscribers.clone(),
+            ref_count: self.ref_count.clone(),
+        }
+    }
+}
+
+impl<S, A> Drop for Protocol<S, A> {
```

Comment:
What happens to the run_loop when the protocol is dropped?

Reply:
It keeps running -- the full shutdown only takes place when both are dropped (which incidentally works in `upstream`). As noted, the run loop does not make much sense anymore, but I want to think about whether we want it to return an error in some cases, or if we can redesign it somehow so we could just restart the network with some altered parameters.

Reply:
Agreed, we should marinate this a bit more before we commit to anything. The only thing worth considering at this point is if the `AbortHandle` that is used in a couple of places now (`testnet` and the upstream proxy) should be the returned by librad.

Reply:
It's not really possible nor meaningful to hand out an `AbortHandle`, or stick some channel receiver into the protocol, _unless_ the protocol would _internally_ drop all the resources (and runaway spawned tasks). But then, what do you do with such an object? You'd need to query it at runtime to see if it is actually running or not.

I'd say async Rust is just inherently anti-modular at its current stage, and RAII the predominant idiom. Up to the library author to make sure `drop` does what it says it does.

Reply:
Understood!